File: 프로그래머스/3/43238. 입국심사/입국심사.cpp
#include <string>
#include <vector>
#define ll long long
using namespace std;

bool chk(int &n, ll &m, vector<int> &times){
    ll cnt = 0;
    for(auto t: times) {
        cnt += m / t;
        if(cnt >= n) return 1;
    }
    return 0;
}

ll solution(int n, vector<int> times) {
    ll l = 0, r = 1e9 * n, m;
    while(l < r){
        m = (l+r)/2;
        if(chk(n,m,times)) r = m;
        else l = m + 1;
    }
    
    return r;
}

File: 프로그래머스/3/43238. 입국심사/입국심사.java
import java.util.*;

class Solution {
    static long MX = 1000000001;
    
    public long solution(int n, int[] times) {
        long l = 0, r = MX * n;
        long m = (l + r) / 2;
        
        while(l <= r) {
            long cnt = 0;
            for(long t : times) cnt += (m / t);
            
            if(cnt >= n) r = m - 1;
            else l = m + 1;
              
            m = (l + r) / 2;
        }
        
        System.out.println("l " + l + " m " + m + " r " + r);
        
        return l;
    }
}